provider/qsd.py

```diff
@@ -0,0 +1,473 @@
+                    lambda: dev.verify_unplug(out, vm.monitor) is True, first=1,
+                    step=5, timeout=20)
+            else:
+                LOG_JOB.info("Ignore device %s Can not be found", dev.get_qid())
+
+
+class QsdDaemonDev(QDaemonDev):
+    # Default data struct of raw image data.
+    raw_image_data = {
+        "protocol": {"driver": "file", "node-name": "tbd", "filename": "tbd",
+                     "auto-read-only": True, "discard": "unmap"},
+        "format": {"driver": "raw", "node-name": "tbd", "file": "tbd"},
+        "filter": {"driver": None, "node-name": "tbd", "file": "tbd"},
+        "name": "",
+        "unix": {"type": "unix", "path": "tbd"},
+        "inet": {"type": "inet", "host": "0.0.0.0"},
+        "nbd-server": {"addr": {}},
+        "export": {"type": "", "id": "", "node-name": "", "writable": True},
+        "image_object": None
+    }
+
+    def __init__(self, name, params):
+        qsd_params = params.object_params(name)
+        basedir = qsd_params.get("qsd_basedir")
+        if not basedir:
+            basedir = data_dir.get_data_dir() + "/qsd/%s" % name
+        if not os.path.exists(basedir):
+            LOG_JOB.info("Create QSD basedir %s", basedir)
+            os.makedirs(basedir)
+
+        binary = qsd_params.get("qsd_binary", "/usr/bin/qemu-storage-daemon")
+        sock_path = qsd_params.get("qsd_sock_path",
+                                   "%s/%s_monitor.sock" % (basedir, name))
+        qsd_monitor_id = "qsd_monitor_%s" % name
+        qid = "qsd_%s" % name
+        super(QsdDaemonDev, self).__init__('QSD', aobject=name,
+                                           child_bus=QUnixSocketBus(qid, qid))
+        self.name = name
+        self.basedir = basedir
+        self.monitor = None
+        self.qsd_params = qsd_params.copy()
+        self.qsd_params["monitor_filename"] = sock_path
+        self.binary = binary
+        self.sock_path = sock_path
+        self.qsd_monitor_id = qsd_monitor_id
+        self.qsd_version = process.run("%s -V" % binary,
+                                       verbose=False,
+                                       ignore_status=True,
+                                       shell=True).stdout_text.split()[2]
+        self.__qsd_help = process.run("%s -h" % binary,
+                                      verbose=False,
+                                      ignore_status=True,
+                                      shell=True).stdout_text
+
+        LOG_JOB.info(self.qsd_version)
+        self.caps = Capabilities()
+        self._probe_capabilities()
+        self.images = {}
+        self.daemonize = False
+        self.pidfile = None
+        self.pid = None
+
+    def _remove_images(self):
+        for img in self.images.values():
+            if img["image_object"]:
+                LOG_JOB.info("QSD ready to remove image:%s", img["name"])
+                if self.qsd_params.get("remove_image_%s" % img["name"],
```

Comment:
1. if users dont want to remove all images after test, they can set a global var remove_image = no, but here, the image could also be removed even remove_image = no.
2. here we only handle the images created by QSD module, do we need to consider the images created by VT? i.e. consider all images in self.images.

Reply:
1. Change to support remove_image = no.
2. To be consistent it is better QSD only handle the images are created by it.